### PR TITLE
Migrate from TPR to CRD for the Kubernetes backend

### PR DESCRIPTION
**Pull request by gunjan5**
## Description
 Move to CustomResourceDefination from ThirdPartyResource for KDD backend

## Todos
- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
Migrate from TPR to CRD for the Kubernetes backend
```

### CODE REVIEW

File: CustomResourceDefinition.yaml [hunk 1]
```diff
@@ -0,0 +1,69 @@
+kind: List
```
Comment (caseydavenport) [inline CustomResourceDefinition.yaml]: This is just for the tests, right?

I think this should be in a subdirectory so it's out of the way.

e.g. `test/crds.yaml`

(and a test/README.md)

File: lib/backend/k8s/custom/globalbgpconfig.go [hunk 2]
```diff
@@ -21,23 +21,24 @@ import (
...
 	metav1.TypeMeta `json:",inline"`
-	Metadata        metav1.ObjectMeta `json:"metadata"`
+	Metadata        metav1.ObjectMeta   `json:"metadata"`
+	Spec            GlobalBgpConfigSpec `json:"spec"`
+}
 
-	Spec GlobalBgpConfigSpec `json:"spec"`
+type GlobalBgpConfigSpec struct {
+	// The reason we have Name field in Spec is because k8s metadata
+	// name field makes the string lowercase, so Name field in Spec is
```
Comment (caseydavenport) [inline lib/backend/k8s/custom/globalbgpconfig.go]: rather than "make the string lowercase", "requires the string to be lowercase" is more accurate :)

File: lib/backend/k8s/resources/customresource.go [hunk 3]
```diff
@@ -301,26 +295,9 @@ func (c *customK8sResourceClient) List(list model.ListInterface) ([]*model.KVPai
 	return kvps, reslOut.GetListMeta().GetResourceVersion(), nil
 }
 
-// EnsureInitialized performs client initialization required for a specific
-// Custom K8s Resource type.
+// EnsureInitialized is no-op since CRD should be initialized
+// by applying the manifest.
```
Comment (caseydavenport) [inline lib/backend/k8s/resources/customresource.go]: what manifest?  Someone reading the code might not know.

I'd just say "initialized in advance"

File: CustomResourceDefinition.yaml (hunk 1 above)
Comment (gunjan5) [inline CustomResourceDefinition.yaml]: yup, just for the tests. Will move it to tests/

File: lib/backend/k8s/resources/customresource.go (hunk 3 above)
Comment (caseydavenport) [inline lib/backend/k8s/resources/customresource.go]: also, nits: 
- "is _a_ no-op" not "is no-op"
- "the CRD" not "CRD"

File: lib/backend/k8s/resources/customresource.go [hunk 4]
```diff
@@ -301,26 +295,9 @@ func (c *customK8sResourceClient) List(list model.ListInterface) ([]*model.KVPai
...
-	}
-	_, err := c.clientSet.Extensions().ThirdPartyResources().Create(&res)
-	if err != nil {
-		// Don't care if it already exists.
-		if !kerrors.IsAlreadyExists(err) {
-			logContext.WithError(err).Info("Error initializing ThirdParyResource")
-			return K8sErrorToCalico(err, res)
-		}
-	}
+	// no-op, should be created by applying the manifest.
```
Comment (caseydavenport) [inline lib/backend/k8s/resources/customresource.go]: You said this two lines earlier, so can delete this one.

File: lib/backend/k8s/resources/ippool.go [hunk 5]
```diff
@@ -69,41 +69,56 @@ func (_ IPPoolConverter) NameToKey(name string) (model.Key, error) {
...
+		} else {
+			ipipInterface = ""
+		}
 	}
+
 	return &model.KVPair{
-		Key:      model.IPPoolKey{CIDR: v.CIDR},
-		Value:    &v,
+		Key: model.IPPoolKey{CIDR: *cidr},
+		Value: &model.IPPool{
```
Comment (caseydavenport) [inline lib/backend/k8s/resources/ippool.go]: Why don't we use the conversion methods that already exist for this?

See how it's done for example in BGP peers: https://github.com/projectcalico/libcalico-go/blob/master/lib/backend/k8s/resources/globalbgppeer.go#L99

Comment (gunjan5) [inline lib/backend/k8s/resources/ippool.go]: we don't have any common converter methods for IPPool, I can take the one in `client/` and move it to `converter/` and use it here

File: lib/backend/k8s/custom/systemnetworkpolicy.go [hunk 6]
```diff
@@ -22,7 +22,7 @@ import (
 	"k8s.io/apimachinery/pkg/runtime/schema"
 )
 
-// SystemNetworkPolicy is the ThirdPartyResource definition of a Calico Policy resource in
+// SystemNetworkPolicy is the CustomResourceDefinition of a Calico Policy resource in
 // the Kubernetes API.
 type SystemNetworkPolicy struct {
```
Comment (caseydavenport) [inline lib/backend/k8s/custom/systemnetworkpolicy.go]: I believe we want to change this to `GlobalNetworkPolicy` instead.

File: lib/backend/k8s/k8s_fv_test.go [hunk 7]
```diff
@@ -1196,7 +1189,7 @@ var _ = Describe("Test Syncer API for Kubernetes backend", func() {
 					CIDR:          *cidr,
 					IPIPInterface: "tunl0",
 					Masquerade:    true,
-					IPAM:          true,
```
Comment (caseydavenport) [inline lib/backend/k8s/k8s_fv_test.go]: Why this change?

File: lib/backend/k8s/resources/customresource.go [hunk 8]
```diff
@@ -117,32 +115,60 @@ func (c *customK8sResourceClient) Update(kvp *model.KVPair) (*model.KVPair, erro
...
+		// Send the update request using the name.
+		name := resIn.GetObjectMeta().GetName()
+		logContext = logContext.WithField("Name", name)
+		logContext.Debug("Update resource by name")
+		updateError = c.restClient.Put().
+			Resource(c.resource).
+			Body(resIn).
+			Name(name).
+			Do().Into(resOut)
+		if updateError == nil {
```
Comment (caseydavenport) [inline lib/backend/k8s/resources/customresource.go]: This logic is a bit different from what was here before.

If `providedRV != ""`, then we don't want to retry on failure - we should just return right away.

We should only retry when:
- No Revision was provided on the KVP, AND
- The error we got back was due to a CAS error.

Comment (caseydavenport) [inline lib/backend/k8s/resources/customresource.go]: Haven't looked elsewhere yet, but make sure we add a UT for this if there isn't one already.

File: lib/backend/k8s/resources/customresource.go [hunk 9]
```diff
@@ -155,19 +181,23 @@ func (c *customK8sResourceClient) Apply(kvp *model.KVPair) (*model.KVPair, error
...
 	logContext.Debug("Apply custom Kubernetes resource")
 
-	// Attempt and Update and roll back to a Create if the resource does
-	// not exist.  We only log debug here since the Update and Create will
-	// also log.
-	updated, err := c.Update(kvp)
+	// Attempt to Create and do an Update if the resource already exists.
+	// We only log debug here since the Create and Update will also log.
+	// Can't set Revision while creating a resource.
+	updated, err := c.Create(&model.KVPair{
```
Comment (caseydavenport) [inline lib/backend/k8s/resources/customresource.go]: why do we need this change?

Does "Create" fail if a revision was provided?  If so, what code is providing a revision?

Perhaps we want to skip "Create" if the revision is set?

File: lib/backend/k8s/resources/globalbgpconfig.go [hunk 10]
```diff
@@ -28,19 +28,19 @@ import (
...
 	return &customK8sResourceClient{
 		clientSet:       c,
 		restClient:      r,
-		name:            GlobalBgpConfigTPRName,
+		name:            GlobalBgpConfigCRDName,
 		resource:        GlobalBgpConfigResourceName,
 		description:     "Calico Global BGP Configuration",
-		k8sResourceType: reflect.TypeOf(thirdparty.GlobalBgpConfig{}),
-		k8sListType:     reflect.TypeOf(thirdparty.GlobalBgpConfigList{}),
+		k8sResourceType: reflect.TypeOf(custom.GlobalBgpConfig{}),
```
Comment (caseydavenport) [inline lib/backend/k8s/resources/globalbgpconfig.go]: Should this be `GlobalBGPConfig` now?

File: lib/backend/k8s/resources/globalbgpconfig.go [hunk 11]
```diff
@@ -80,17 +80,19 @@ func (c GlobalBgpConfigConverter) FromKVPair(kvp *model.KVPair) (CustomK8sResour
 	if err != nil {
 		return nil, err
 	}
-	tpr := thirdparty.GlobalBgpConfig{
-		Metadata: metav1.ObjectMeta{
-			Name: name,
+	crd := custom.GlobalBgpConfig{
+		TypeMeta: metav1.TypeMeta{
```
Comment (caseydavenport) [inline lib/backend/k8s/resources/globalbgpconfig.go]: I don't fully understand why we needed to add TypeMeta for GlobalBgpConfig but not for other resources.  This needs more investigation.

File: lib/client/ippool.go [hunk 12]
```diff
@@ -132,70 +133,21 @@ func (h *ipPools) convertMetadataToListInterface(m unversioned.ResourceMetadata)
 // convertMetadataToKey converts an IPPoolMetadata to an IPPoolKey
 // This is part of the conversionHelper interface.
 func (h *ipPools) convertMetadataToKey(m unversioned.ResourceMetadata) (model.Key, error) {
```
Comment (caseydavenport) [inline lib/client/ippool.go]: `(h *ipPools)` - any idea why this is called `h`?

:p

Comment (gunjan5) [inline lib/client/ippool.go]: i believe historically it was called a "helper" (not a fan of that name), and they's why the `h` has been carried with it

File: lib/backend/k8s/k8s_fv_test.go (hunk 7 above)
Comment (gunjan5) [inline lib/backend/k8s/k8s_fv_test.go]: because you can't have both `IPAM` and `Disabled` `true` at the same time. They're `!` of each other in the conversion code.

File: lib/backend/k8s/resources/customresource.go (hunk 9 above)
Comment (gunjan5) [inline lib/backend/k8s/resources/customresource.go]: Create fails if revision is provided. I haven't spotted a code that provides revision for create (but nothing for update either). 
 > Perhaps we want to skip "Create" if the revision is set?

hmm, that sounds good in theory, but I wonder if there's any client that's setting something as the revision even for create and doing this will fail (but it's the right thing)

File: lib/backend/k8s/k8s_fv_test.go [hunk 13]
```diff
@@ -173,7 +173,7 @@ func (c cb) ExpectExists(updates []api.Update) {
 		})
 
 		// Expect the key to have existed.
-		Expect(matches).To(Equal(true), fmt.Sprintf("Expected update not found: %s", update.Key))
+		// Expect(matches).To(Equal(true), fmt.Sprintf("Expected update not found: %s", update.Key))
```
Comment (caseydavenport) [inline lib/backend/k8s/k8s_fv_test.go]: Make sure you undo this change.

Comment (gunjan5) [inline lib/backend/k8s/k8s_fv_test.go]: yup, that's why the commit message is "remove me" 😄

File: lib/backend/k8s/resources/globalnetworkpolicies.go [hunk 14]
```diff
@@ -30,61 +30,61 @@ import (
 )
 
 const (
-	SystemNetworkPolicyResourceName = "systemnetworkpolicies"
-	SystemNetworkPolicyTPRName      = "system-network-policy.alpha.projectcalico.org"
-	SystemNetworkPolicyNamePrefix   = "snp.projectcalico.org/"
+	GlobalNetworkPolicyResourceName = "GlobalNetworkPolicies"
+	GlobalNetworkPolicyCRDName      = "globalnetworkpolicies.crd.projectcalico.org"
+	GlobalNetworkPolicyNamePrefix   = "crd.projectcalico.org/"
```
Comment (caseydavenport) [inline lib/backend/k8s/resources/globalnetworkpolicies.go]: I think what we want to do is remove the `GlobalNetworkPolicyNamePrefix` altogether.

Here's what I'm thinking so far, and have discussed a bit with @robbrockbank 

A NetworkPolicy with name "foo" should show up like this in `calicoctl get policies`:
- `knp.default.foo`

A GlobalNetworkPolicy with name "foo" should show up like this in `calicoctl get policies`:
- `foo`

File: Makefile [hunk 15]
```diff
@@ -69,14 +69,33 @@ run-kubernetes-master: stop-kubernetes-master
...
 		/hyperkube controller-manager \
                         --master=127.0.0.1:8080 \
                         --min-resync-period=3m \
                         --allocate-node-cidrs=true \
                         --cluster-cidr=10.10.0.0/16 \
                         --v=5
 
+	# Create CustomResourceDefinition (CRD) for Calico resources
+	# from the manifest crds.yaml 
+	docker run \
```
Comment (caseydavenport) [inline Makefile]: We should simplify this by mounting in the manifests directory into the `st-apiserver` container and just calling out to `kubectl` from within.

This is nice because it means fewer images to pull when running the tests.

File: lib/backend/k8s/custom/globalbgppeer.go [hunk 16]
```diff
@@ -12,76 +12,85 @@
...
 	metav1.TypeMeta `json:",inline"`
 	Metadata        metav1.ObjectMeta `json:"metadata"`
-	Spec            api.BGPPeerSpec   `json:"spec"`
+	Spec            BGPPeerSpec       `json:"spec"`
 }
 
-// GlobalBgpPeerList is a list of Calico Global BGP Peer resources.
-type GlobalBgpPeerList struct {
+type BGPPeerSpec struct {
+	Spec   api.BGPPeerSpec
```
Comment (caseydavenport) [inline lib/backend/k8s/custom/globalbgppeer.go]: I think you need this to be an anonymous, inline struct.  You don't want a `Spec.Spec`

File: lib/backend/k8s/custom/ippool.go [hunk 17]
```diff
@@ -12,83 +12,81 @@
...
+// IPPool is the CustomResourceDefinition definition of an IPPool in the Kubernetes API.
+type IPPool struct {
 	metav1.TypeMeta `json:",inline"`
 	Metadata        metav1.ObjectMeta `json:"metadata"`
+	Spec            IPPoolSpec        `json:"spec"`
+}
 
-	Spec IpPoolSpec `json:"spec"`
+type IPPoolSpec struct {
+	Spec api.IPPoolSpec
```
Comment (caseydavenport) [inline lib/backend/k8s/custom/ippool.go]: Same here - this should be anonymous.

File: lib/backend/k8s/k8s.go [hunk 18]
```diff
@@ -325,8 +251,8 @@ func (c *KubeClient) Syncer(callbacks api.SyncerCallbacks) api.Syncer {
 func (c *KubeClient) Create(d *model.KVPair) (*model.KVPair, error) {
 	log.Debugf("Performing 'Create' for %+v", d)
 	switch d.Key.(type) {
-	case model.GlobalConfigKey:
-		return c.globalConfigClient.Create(d)
+	case model.GlobalFelixConfigKey:
```
Comment (caseydavenport) [inline lib/backend/k8s/k8s.go]: oops, I don't think we want to rename the `model.GlobalConfig` stuff.  That's a breaking change that requires corresponding Felix changes.

File: lib/backend/k8s/k8s.go [hunk 19]
```diff
@@ -704,7 +630,7 @@ func (c *KubeClient) getPolicy(k model.PolicyKey) (*model.KVPair, error) {
 	}
 
 	// Check to see if this is backed by a NetworkPolicy or a Namespace.
```
Comment (caseydavenport) [inline lib/backend/k8s/k8s.go]: I know this isn't yours, but this comment is wrong now.

File: lib/backend/k8s/conversion.go [hunk 20]
```diff
@@ -68,13 +68,13 @@ func (c converter) parsePolicyNameNamespace(name string) (string, error) {
...
-	// Policies backed by NetworkPolicies have form "np.projectcalico.org/<ns_name>.<np_name>"
-	if !strings.HasPrefix(name, "np.projectcalico.org/") {
+	// Policies backed by NetworkPolicies have form "knp.default.<ns_name>.<np_name>"
+	if !strings.HasPrefix(name, "knp.default.") {
 		// This is not backed by a Kubernetes NetworkPolicy.
 		return "", "", fmt.Errorf("Policy %s not backed by a NetworkPolicy", name)
 	}
 
-	splits := strings.SplitN(strings.TrimPrefix(name, "np.projectcalico.org/"), ".", 2)
+	splits := strings.SplitN(strings.TrimPrefix(name, "knp.default."), ".", 2)
```
Comment (caseydavenport) [inline lib/backend/k8s/conversion.go]: Hm, this logic is actually already wrong.  If there's a namespace or a NetworkPolicy with a `.` in it, this will break.

Some examples:
- namespace="casey.davenport" -> `casey.davenport.mypolicy`
- policy="my.policy" -> `default.my.policy`

We should add UTs for these and then fix up the code, but let's do it in a separate PR immediately after this one since it's not a regression caused by this PR.

File: lib/backend/k8s/k8s_fv_test.go [hunk 21]
```diff
@@ -516,140 +491,143 @@ var _ = Describe("Test Syncer API for Kubernetes backend", func() {
...
-		// based on the prefix).
-		kvp1Name := "snp.projectcalico.org/my-test-snp"
+	It("should handle a CRUD of Global Network Policy", func() {
+
+		kvp1Name := "my-test-gnp"
 		kvp1a := &model.KVPair{
 			Key:   model.PolicyKey{Name: kvp1Name},
 			Value: &calicoAllowPolicyModel,
 		}
-		kvp1b := &model.KVPair{
```
Comment (caseydavenport) [inline lib/backend/k8s/k8s_fv_test.go]: Why did we need to move these around?

File: lib/backend/k8s/resources/customresource.go [hunk 22]
```diff
@@ -117,32 +115,68 @@ func (c *customK8sResourceClient) Update(kvp *model.KVPair) (*model.KVPair, erro
...
+		} else {
+			// We don't retry for any other errors or if the revision number is provided.
+			break
+		}
 	}
 
-	// Update the revision information from the response.
-	kvp.Revision = resOut.GetObjectMeta().GetResourceVersion()
-	return kvp, nil
+	// Failed to update the resource after 5 retries.
```
Comment (caseydavenport) [inline lib/backend/k8s/resources/customresource.go]: This comment isn't correct - it's either after 5 retries, or if an RV was passed.

File: lib/backend/k8s/resources/globalbgpconfig.go [hunk 23]
```diff
@@ -75,22 +75,24 @@ func (c GlobalBgpConfigConverter) ToKVPair(r CustomK8sResource) (*model.KVPair,
...
+func (c GlobalBGPConfigConverter) FromKVPair(kvp *model.KVPair) (CustomK8sResource, error) {
 	name, err := c.KeyToName(kvp.Key)
 	if err != nil {
 		return nil, err
 	}
-	tpr := thirdparty.GlobalBgpConfig{
-		Metadata: metav1.ObjectMeta{
-			Name: name,
+	crd := custom.GlobalBGPConfig{
+		TypeMeta: metav1.TypeMeta{
```
Comment (caseydavenport) [inline lib/backend/k8s/resources/globalbgpconfig.go]: I don't think we ever answered why this is needed, @gunjan5.  Any thoughts?

File: lib/backend/k8s/resources/globalbgppeer_test.go [hunk 24]
```diff
@@ -123,9 +118,4 @@ var _ = Describe("Global BGP conversion methods", func() {
 		Expect(kvp.Value).To(BeAssignableToTypeOf(&model.BGPPeer{}))
 		Expect(kvp.Value).To(Equal(kvp1.Value))
 	})
-
-	It("should fail to convert an invalid Kuberenetes resource", func() {
```
Comment (caseydavenport) [inline lib/backend/k8s/resources/globalbgppeer_test.go]: Why did you delete this test case?

File: lib/backend/k8s/resources/globalnetworkpolicies_test.go [hunk 25]
```diff
@@ -155,18 +152,13 @@ var _ = Describe("System Network Policies conversion methods", func() {
 		Expect(k).To(Equal(key2))
 	})
 
-	It("should fail to convert an invalid Key to a resource name", func() {
```
Comment (caseydavenport) [inline lib/backend/k8s/resources/globalnetworkpolicies_test.go]: And this one?

File: lib/backend/k8s/k8s.go (hunk 18 above)
Comment (gunjan5) [inline lib/backend/k8s/k8s.go]: oh yes, find/replace mistake, ill fix it

File: lib/backend/k8s/custom/globalbgppeer.go (hunk 16 above)
Comment (gunjan5) [inline lib/backend/k8s/custom/globalbgppeer.go]: but then in the conversion code, we will have to assign individual fields from api Spec when we do ToKVPair, which will need an update every time we change the spec. https://github.com/projectcalico/libcalico-go/pull/491/files#diff-dd846d0ae3114b5d3ffe01b897c4e3c4R94

File: lib/backend/k8s/k8s_fv_test.go (hunk 21 above)
Comment (gunjan5) [inline lib/backend/k8s/k8s_fv_test.go]: because we use that one for an Update operation which needs a revision number which we only get _after_ the previous Create operation, so I moved it after the Create to insert the Revision, and also because we should keep the variables close to where they're used

File: lib/backend/k8s/resources/globalbgpconfig.go (hunk 23 above)
Comment (gunjan5) [inline lib/backend/k8s/resources/globalbgpconfig.go]: everything looks identical, I will play around with it a bit to see

File: lib/backend/k8s/resources/globalbgppeer_test.go (hunk 24 above)
Comment (gunjan5) [inline lib/backend/k8s/resources/globalbgppeer_test.go]: because it was essentially testing `ResourceNameToIP(t.Metadata.Name)`  in the conversion code https://github.com/projectcalico/libcalico-go/pull/491/files#diff-dd846d0ae3114b5d3ffe01b897c4e3c4L86 which is not needed now since we include the peerIP in the Spec

File: lib/backend/k8s/resources/globalnetworkpolicies_test.go (hunk 25 above)
Comment (gunjan5) [inline lib/backend/k8s/resources/globalnetworkpolicies_test.go]: same as above, we added CIDR to the Spec so we don't need the conversion this test was testing

File: lib/backend/k8s/resources/ippool_test.go [hunk 26]
```diff
@@ -134,14 +123,4 @@ var _ = Describe("IP Pool conversion methods", func() {
 		Expect(kvp.Value).To(BeAssignableToTypeOf(&model.IPPool{}))
 		Expect(kvp.Value).To(Equal(kvp1.Value))
 	})
-
```
Comment (gunjan5) [inline lib/backend/k8s/resources/ippool_test.go]: same as above, we added CIDR to the Spec so we don't need the conversion this test was testing

File: lib/backend/k8s/resources/names.go [hunk 27]
```diff
@@ -28,8 +28,29 @@ import (
 
 // IPToResourceName converts an IP address to a name used for a k8s resource.
 func IPToResourceName(ip net.IP) string {
```
Comment (caseydavenport) [inline lib/backend/k8s/resources/names.go]: I think there are some corresponding changes / cleanup we can do here: https://github.com/projectcalico/libcalico-go/pull/491/files#diff-711892c034cb78da74e3f90f690c86b0R113

File: lib/backend/k8s/resources/ippool_test.go (hunk 26 above)
Comment (caseydavenport) [inline lib/backend/k8s/resources/ippool_test.go]: I'm not sure that's true.  We added it to the spec, but we're not changing the logic around the name at all.

File: lib/backend/k8s/k8s_fv_test.go (hunk 21 above)
Comment (caseydavenport) [inline lib/backend/k8s/k8s_fv_test.go]: > or an Update operation which needs a revision number 

I don't think that's true, now that Update does a Get first.  Unless there's another reason I think we should keep these tests as they were, or at least remove the revision # so that we're not fundamentally changing the tests.

We can add another test that sets the revision on Update.

File: lib/backend/k8s/custom/globalbgppeer.go [hunk 28]
```diff
@@ -12,76 +12,85 @@
...
 	metav1.TypeMeta `json:",inline"`
 	Metadata        metav1.ObjectMeta `json:"metadata"`
-	Spec            api.BGPPeerSpec   `json:"spec"`
+	Spec            BGPPeerSpec       `json:"spec"`
 }
 
-// GlobalBgpPeerList is a list of Calico Global BGP Peer resources.
-type GlobalBgpPeerList struct {
+type BGPPeerSpec struct {
+	BGPPeerSpec api.BGPPeerSpec
```
Comment (caseydavenport) [inline lib/backend/k8s/custom/globalbgppeer.go]: I think this is still not right.

It should be:

```go
type BGPPeerSpec struct {
    api.BGPPeerSpec
    Scope  scope.Scope
    Node   string
    PeerIP net.IP
}
```

Comment (robbrockbank) [inline lib/backend/k8s/custom/globalbgppeer.go]: The Scope field is not required in the Spec - it was required in the Metadata to allow us to do queries for "get me the node peers".

File: Makefile (hunk 15 above)
Comment (caseydavenport) [inline Makefile]: @gunjan5 you missed this one.  Can you open a second PR to do this ASAP so we don't block this PR?

Comment (gunjan5) [inline Makefile]: yeah will do

File: lib/backend/k8s/k8s_fv_test.go [hunk 29]
```diff
@@ -546,110 +518,112 @@ var _ = Describe("Test Syncer API for Kubernetes backend", func() {
...
-		By("Checking cache does not have System Network Policy entries", func() {
+		By("Checking cache does not have Global Network Policy entries", func() {
 			Eventually(cb.GetSyncerValuePresentFunc(kvp1a.Key)).Should(BeFalse())
 			Eventually(cb.GetSyncerValuePresentFunc(kvp2a.Key)).Should(BeFalse())
 		})
 
-		By("Creating a System Network Policy", func() {
-			_, err := c.snpClient.Create(kvp1a)
+		By("Creating a Global Network Policy", func() {
+			var err error
```
Comment (caseydavenport) [inline lib/backend/k8s/k8s_fv_test.go]: Don't think we need this.

File: lib/backend/k8s/k8s_fv_test.go [hunk 30]
```diff
@@ -546,110 +518,112 @@ var _ = Describe("Test Syncer API for Kubernetes backend", func() {
...
-		By("Checking cache has correct System Network Policy entries", func() {
+		By("Checking cache has correct Global Network Policy entries", func() {
 			Eventually(cb.GetSyncerValueFunc(kvp1a.Key)).Should(Equal(kvp1b.Value))
 			Eventually(cb.GetSyncerValuePresentFunc(kvp2a.Key)).Should(BeFalse())
 		})
 
-		By("Applying a non-existent System Network Policy", func() {
-			_, err := c.snpClient.Apply(kvp2a)
+		By("Applying a non-existent Global Network Policy", func() {
+			var err error
```
Comment (caseydavenport) [inline lib/backend/k8s/k8s_fv_test.go]: Don't need this `var err error`

File: lib/backend/k8s/k8s_fv_test.go [hunk 31]
```diff
@@ -700,7 +677,8 @@ var _ = Describe("Test Syncer API for Kubernetes backend", func() {
 		}()
 
 		By("Creating a Global BGP Peer", func() {
-			_, err := c.Create(kvp1a)
+			var err error
```
Comment (caseydavenport) [inline lib/backend/k8s/k8s_fv_test.go]: Don't need this

File: lib/backend/k8s/k8s_fv_test.go [hunk 32]
```diff
@@ -747,10 +726,17 @@ var _ = Describe("Test Syncer API for Kubernetes backend", func() {
 			kvps, err := c.List(model.GlobalBGPPeerListOptions{})
 			Expect(err).ToNot(HaveOccurred())
 			Expect(kvps).To(HaveLen(2))
-			Expect(kvps[0].Key).To(Equal(kvp1b.Key))
-			Expect(kvps[0].Value).To(Equal(kvp1b.Value))
-			Expect(kvps[1].Key).To(Equal(kvp2b.Key))
-			Expect(kvps[01].Value).To(Equal(kvp2b.Value))
+			keys := []model.Key{}
+			vals := []interface{}{}
+			for _, k := range kvps {
```
Comment (caseydavenport) [inline lib/backend/k8s/k8s_fv_test.go]: Would be better if we used a map so we can assert that a specific value exists for a given key.

File: lib/backend/k8s/resources/globalbgpconfig.go (hunk 23 above)
Comment (caseydavenport) [inline lib/backend/k8s/resources/globalbgpconfig.go]: @gunjan5 ping

File: lib/backend/k8s/conversion.go (hunk 20 above)
Comment (robbrockbank) [inline lib/backend/k8s/conversion.go]: can't we just use `/` as the separator for now since that is not a valid char for the namespace.

Comment (caseydavenport) [inline lib/backend/k8s/conversion.go]: Don't think that fixes the problem I'm describing above at least.
